b2g/components/RemoteControlService.jsm

```diff
@@ -491,30 +624,342 @@ function getCommandJSSandbox(aImportFunctions) {
 // Parse and dispatch incoming events from client
 function EventHandler(aConnection) {
   this._connection = aConnection;
+  this._server = aConnection.server;
 
   aConnection.eventHandler = this;
+
+  this._status = EVENT_HANDLER_STATUS.INITIAL;
+
+  this._clientID = null;
+  this._aes256B64 = {};
+  this._hmac256B64 = {};
+  this._hmac256Key = null;
+
+  this._JPAKE = null;
+  // For round 1
+  this._gx1 = {};
+  this._gv1 = {};
+  this._r1 = {};
+  this._gx2 = {};
+  this._gv2 = {};
+  this._r2 = {};
+  // For round 2
+  this._A = {};
+  this._gva = {};
+  this._ra = {};
+
+  this._client_round1 = null;
+  this._client_round2 = null;
 }
 EventHandler.prototype = {
   // PUBLIC FUNCTIONS
   handleEvent: function(aEvent) {
-    // TODO: Implement JPAKE pairing (Bug 1207996)
     switch (aEvent.type) {
-      // Implement control command dispatch
+      case "auth":
+        // Implement JPAKE authentication
+        this._handleAuthEvent(aEvent);
+        break;
       case "command":
+        // Implement control command dispatch
         this._handleCommandEvent(aEvent);
         break;
       default:
+        // Not accepted event type, report error to client
+        this._handleError(this, "common", "Not accepted event type");
+        break;
+    }
+  },
+
+  _handleAuthEvent: function(aEvent) {
+    switch (aEvent.action) {
+      case "request_handshake":
+        this._handleHandshake(aEvent);
+        break;
+      case "jpake_client_1":
+        this._handleJPAKERound1Event(aEvent);
+        break;
+      case "jpake_client_2":
+        this._handleJPAKERound2Event(aEvent);
+        break;
+      case "client_key_confirmation":
+        this._handleKeyConfirmEvent(aEvent);
+        break;
+      default:
         break;
     }
   },
 
-  // Pass and run command event in sandbox
+  _handleHandshake: function(aEvent) {
+    if (this._status != EVENT_HANDLER_STATUS.INITIAL) {
+      DEBUG && debug("Status not in INITIAL, drop the event");
+      this._handleError(this, aEvent.type, "Wrong status");
+      return;
+    }
+
+    let reply = {
+      type: "auth",
+      action: "response_handshake",
+      detail: 1
+    };
+
+    // If client sends an existing ID, then reply is 2 (second connection)
+    // Also restore AES and HMAC key from UUID
+    if (aEvent.detail && aEvent.detail.id && this._server.getKeysFromUUID(aEvent.detail.id) !== null) {
+      reply.detail = 2;
+
+      this._clientID = aEvent.detail.id;
+      let keys = this._server.getKeysFromUUID(aEvent.detail.id);
+      this._aes256B64 = { value: keys.aesKey };
+      this._hmac256B64 = { value: keys.hmacKey };
+    }
+
+    this._connection.sendString(JSON.stringify(reply));
```

Comment:
looks like you always use `sendString` with a JSON string, why not just call it `sendMsg` with JSON string conversion and pass the JSON object directly.